---
title: "RNAseq Analysis"
author: "Jordan Prahl"
date: "7/7/2020"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
BiocManager::install("edgeR")
BiocManager::install("limma")
BiocManager::install("AnnotationDbi")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("umap")
library(edgeR)
library(dplyr)
library(tidyverse)
library(limma)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(reshape2)
library(ggplot2)
library(ggfortify)
library(magrittr)
library(tibble)
library(DataCombine)
library(umap)
```

load count files from STAR and put into a matrix 
```{r}
ff <- list.files( path = "./readCounts", pattern = "*ReadsPerGene.out.tab$", full.names = TRUE ) # make a list of file names
counts.files <- lapply( ff, read.table, skip = 4 ) ### read files and create data frame without first 4 rows 
counts <- sapply( counts.files, function(x) x[ , 2 ] ) ### put counts for each sample into a matrix based on values in the second column (non-stranded counts)
ff <- gsub( "_ReadsPerGene.out.tab", "", ff ) ### remove "_ReadsPerGene.out.tab" from end of file names
ff <- gsub( "./readCounts", "", ff ) ### remove file path from sample names
colnames(counts) <- ff ### add column lables to count matrix (sample names)
row.names(counts) <- counts.files[[1]]$V1 ### add the ensemble ID to rows of count matrix
```

remove decimals from ENSEMBLE IDs
```{r}
rownames(counts) <- gsub("\\..+","",rownames(counts))
head(counts)
```


remove duplicates from ENSEMBLE IDs
```{r}
counts <- unique.matrix(counts)
head(counts)
summary(counts)
```

create a DGEList from counts
```{r}
my_dge <- DGEList(counts, genes=rownames(counts))

# replace org.Mm.eg.db with org.Hs.eg.db
my_dge$genes$Symbol <- mapIds(org.Hs.eg.db,
                         rownames(my_dge),
                         keytype="ENSEMBL", column="SYMBOL")
```

filter counts
```{r}
keep <- rowSums(cpm(my_dge) > 0.5) >= 2
table(keep)
```


```{r}
my_dge <- my_dge[keep, , keep.lib.sizes=FALSE]
```

normalize the counts
```{r}
my_dge <- calcNormFactors(my_dge)
```

calculate CPM and join gene annotation to dataframe
```{r}

expr_data <- data.frame(cpm(my_dge)) %>% 
  rownames_to_column(var="genes") %>%
  dplyr::left_join(data.frame(my_dge$genes), by="genes") %>%
  dplyr::select(genes,Symbol,everything())
                   
write.csv(data.frame(expr_data), './annotations_new.csv')

summary(expr_data)

head(expr_data)
```

```{r, read in annotation data}
expr_data <- read.csv('./annotations_new.csv')
```

```{r, melt data}
m.data <- melt(expr_data)

Replaces <- data.frame(from = c("X.JPA1", "X.JPA2", "X.JPA3", "X.JPG1", "X.JPG2", "X.JPG3", "X.JPW1", "X.JPW2", "X.JPW3"), to = c("A/-", "A/-", "A/-", "G/-", "G/-", "G/-", "A/G", "A/G", "A/G"))


r.m.data <- FindReplace(data = m.data, Var = "variable", replaceData = Replaces,
                     from = "from", to = "to", exact = FALSE)

head(r.m.data)
```

```{r, single gene expression}
GENE <- "BAG3"

gdata1 <- subset(gdata, gdata[,"Symbol"]==GENE)

model1 <- aov(data = gdata1, value~variable)
text1 <- summary(model1)
text2 <- text1[[1]]$'Pr(>F)'

g <- ggplot(gdata1, aes(variable,value))
g + geom_boxplot(fill='#A4A4A4', color="black") +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1.0) +
  theme_minimal() +
  ylab("Read Counts") + xlab("Genotype") + ggtitle(GENE, text2) + 
  theme(
    plot.title = element_text(hjust = 0.5, size=25, face="bold"), 
    axis.text.x = element_text(size=15),
    axis.text.y = element_text(size=12)
  ) 
```


```{r, Dimension Reduction Data Prep}
#Reading in the full data from Steve's analysis
BiocManager::install("ggpubr")
BiocManager::install("Rtsne")
BiocManager::install("M3C")
library(readxl)
library(M3C)
library(Rtsne)
library(umap)
library(ggpubr)

full_data <- read_excel("202000903_7compar_luhmes_lesion_jp_Results_all.xlsx")
View(full_data)
# filter the data
full_data <- full_data[, grep("^logFC", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^AveExpr", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^t_", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^P.Value", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^adj.P", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^B_", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^name", colnames(full_data), invert = T)]
full_data <- full_data[, grep("^SNCA", colnames(full_data), invert = T)]
ids <- full_data[,1]
geneNames <- full_data[,1:2]
full_data <- full_data[, grep("^symbol", colnames(full_data), invert = T)]
just_data <- full_data[, grep("^ensembl", colnames(full_data), invert = T)]

# For UMAP we need the samples in the rows and genes in columns
df2 <- t(just_data)
```

```{r, PCA}
pca_data <- prcomp(t(just_data), scale. = T)

pData <- data.frame(pca_data$x)
pData$Geno <- c("WT", "WT","WT", "WT","WT", "WT","WT", "WT","WT", "WT","CRISPR Contol","CRISPR Contol","CRISPR Contol","CRISPR Contol","CRISPR Contol","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","A/-","A/-","A/-","G/-","G/-","G/-","A/G","A/G","A/G")
View(pData)

write.csv2(data.frame(pData), './PCA_dataframe.csv')

barplot(pca_data$sdev) # Bar graph of variance
summary(pca_data)

var_explained <- pca_data$sdev^2/sum(pca_data$sdev^2) # Add variance to plot axis
my_comparisons <- list(c("A/-", "G/-"), c("A/-", "Les Control"), c("G/-", "Les Control")) # Pairwise comparisons for stats

pData1.1 <- subset.data.frame(pData[25:33,])
View(pData1.1)

p23 <- ggplot(data= pData, aes(x=PC2, y=PC3)) +
  geom_point(aes(color=Geno, size = 10)) + 
  ggtitle("PCA of RNAseq data") +
  xlab(paste0("PC2: ",round(var_explained[2]*100,1),"% variance explained")) + 
  ylab(paste0("PC3: ",round(var_explained[3]*100,1),"% variance explained"))

p


ggsave("PCA_figure_1_2.png", width = 10, height = 5)

### Box and whisker plot of dimensions
q2 <- ggplot(data=pData1.1, aes(x=Geno, y=PC2)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons, label = "p.signif", paired = F) +
  #stat_compare_means(method = "anova", label.y = 50, label.x=3) +
    theme_minimal()

q2

```

``` {r, UMAP}
# Next, run umap to reduce dimensionality of data
rna.umap <- umap::umap(t(just_data), n_components=5, scale=T)
rna.umap #for summary of data

uData <- data.frame(rna.umap$layout)

uData$Geno <- c("WT", "WT","WT", "WT","WT", "WT","WT", "WT","WT", "WT","CRISPR Contol","CRISPR Contol","CRISPR Contol","CRISPR Contol","CRISPR Contol","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","A/-","A/-","A/-","G/-","G/-","G/-","A/G","A/G","A/G")

colnames(uData) <- c("PC1","PC2","PC3","PC4","PC5","Geno")

write.csv2(data.frame(uData), './UMAP_dataframe.csv')


```
 
```{r, t-SNE}

tsne <- Rtsne(t(dat), perplexity = 5, scale=T, dims = 3) # got warning perplexity is too large

dat2 <- tsne$Y
Geno <-c("WT", "WT","WT", "WT","WT", "WT","WT", "WT","WT", "WT","CRISPR Contol","CRISPR Contol","CRISPR Contol","CRISPR Contol","CRISPR Contol","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","Undiff","A/-","A/-","A/-","G/-","G/-","G/-","A/G","A/G","A/G")

tData <-cbind.data.frame(Geno, dat2)
head(tData)
colnames(tData) <- c("Geno", "PC1", "PC2", "PC3")

write.csv2(data.frame(tData), './TSNE_dataframe.csv')
```



```{r, DIMENSIONALITY PLOTS}

{ #START PLOT HERE
     # Choose your dimensions in the xydims vector below.
     # PCA has 33 dimensions, UMAP has 5, and TSNE only has 3 dimensions
  
  xydims <- c(
              "PC2",  # X-Axis
              "PC3",  # Y-axis
              "TSNE"  # Dimension reduction method: UMAP, PCA, or TSNE
              ) 
  
  if(xydims[3]=="UMAP") {data1=uData} else if(xydims[3]=="PCA") {data1=pData} else {data1=tData}
  
  ggplot(data=data1, aes_string(x=noquote(paste0(xydims[1])),y=noquote(paste0(xydims[2])))) +
    geom_point(aes(color = Geno, size=10)) +
    #geom_density2d(alpha=.5) +
    ggtitle(paste0(xydims[3]," of RNAseq data")) +
    xlab(paste0(xydims[1])) + 
    ylab(paste0(xydims[2])) +
    theme_minimal() +
    scale_colour_manual(values=cbPalette)
  
  ggsave(paste0("./plots/",xydims[3],"_figure_",xydims[1],"_",xydims[2],".png"), width = 10, height = 5)

}

```
